File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if (head == nullptr || head->next == nullptr) return true;

        // Step 1: Find middle
        ListNode* slow = head;
        ListNode* fast = head;

        while (fast->next != nullptr && fast->next->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
        }

        // Step 2: Reverse second half
        ListNode* newHead = reverseList(slow->next);

        // Step 3: Compare both halves
        ListNode* first = head;
        ListNode* second = newHead;

        while (second != nullptr) {
            if (first->val != second->val) {
                reverseList(newHead); // Optional: Restore original structure
                return false;
            }
            first = first->next;
            second = second->next;
        }

        reverseList(newHead); // Optional
        return true;
    }

private:
    // Reverse linked list function
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        while (head != nullptr) {
            ListNode* nextTemp = head->next;
            head->next = prev;
            prev = head;
            head = nextTemp;
        }
        return prev;
    }
};

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {

    private ListNode reverseList(ListNode head) {
    ListNode prev = null;
    while (head != null) {
        ListNode next = head.next;
        head.next = prev;
        prev = head;
        head = next;
    }
    return prev;
}


    public boolean isPalindrome(ListNode head) {
        
    ListNode slow = head;
    ListNode  fast = head;

    if(head==null || head.next==null) return true;

    while(fast.next!=null && fast.next.next!=null){
        slow=slow.next;
        fast=fast.next.next;
    }

    ListNode newhead=reverseList(slow.next);
    ListNode first=head;
    ListNode second=newhead;

    while(second!= null){
        
        if(first.val!=second.val){
            reverseList(newhead);
            return false;
        }
        first=first.next;
        second=second.next;

    }
        return true;
    }
}